### PR TITLE
Zap ci pending changes

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

While changing something into `src/app/zap-templates` or any of the `.zap` files, one needs to run `scripts/tools/zap_regen_all.py` in order to update the tree with those changes.
This step is forgotten very often, and while working on some zap related issues there is very often changes unrelated to what I'm working on.

This patch runs `scripts/tools/zap_regen_all.py` and then `git diff-index --quiet HEAD --` in CI to make sure there is no pending changes.

For sanity checking, I'm expecting this PR to fail the ZAP CI part since there is some pending ZAP changes in the tree...

 #### Summary of Changes
 * Replace the various `scripts/tools/zap_generate*.sh` files by a single python script
 * CI check for pending changes

### CODE REVIEW

File: .github/workflows/zap_templates.yaml
Comment (Damian-Nordic) [inline .github/workflows/zap_templates.yaml:50]: This will only compare files already tracked by git, right? Should `zap_regen_all.py` have an option to automatically `git add` the generated files to make the workflow detect new ZAPT files products of which were not committed?

Comment (vivien-apple) [reply .github/workflows/zap_templates.yaml:50]: Good point. I guess I forgot about new files.  I should probably use `git status` instead of `git diff`. Will look at that.

Comment (vivien-apple) [reply .github/workflows/zap_templates.yaml]: I end up using `git add .` as you first suggest.

File: scripts/tools/zap/generate.py
Comment (andy31415) [inline scripts/tools/zap/generate.py:62]: does isfile not imply exists? (same for isdir below)?

File: scripts/tools/zap/generate.py
Comment (andy31415) [inline scripts/tools/zap/generate.py:28]: could you use the argparse module instead? provides nice --help and more extensible

File: scripts/tools/zap/generate.py
Comment (andy31415) [inline scripts/tools/zap/generate.py:75]: consider adding quotes.
Can also use subprocess and use list arrays and not worry about quoting.

**Review by Damian-Nordic — APPROVED** [review]

**Comment by vivien-apple** [comment]
I pushed an update, with some documentation updates into `src/app/zap-templates/README.md` The documentation there is pretty old.

**Comment by vivien-apple** [comment]
FTR I also opened https://github.com/project-chip/zap/pull/123 against ZAP to make sure `endpoint_config.h` does not have random changes when attributes are moved around internally.

Also, I'm letting this PR with the ZAP CI failure until the last minute. When it will have all approvals I will update the `gen/` folders hoping this is the last time they are out of sync.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by mspang — APPROVED** [review]

File: src/app/zap-templates/README.md
Comment (LuDuda) [inline src/app/zap-templates/README.md:61]:
```suggestion
For example, to generate `chip-tool` additional files the command is:
```

File: src/app/zap-templates/README.md
Comment (LuDuda) [inline src/app/zap-templates/README.md:20]:
```suggestion
need to be updated. You can regenerate all files with:
```

File: src/app/zap-templates/README.md
Comment (LuDuda) [inline src/app/zap-templates/README.md:28]:
```suggestion
By default generated files are located in a `gen/` folder next to the
```

**Review by LuDuda — APPROVED** [review]
This is great, thanks :+1:

**Comment by woody-apple** [comment]
@vivien-apple can you take a peek at this conflict? Thanks!

**Comment by vivien-apple** [comment]
> @vivien-apple can you take a peek at this conflict? Thanks!

I will rebased it ASAP.

**Comment by mspang** [comment]
I looked at this a few days ago and ZAP still seemed to be nondeterministic. Is it fixed?

**Comment by woody-apple** [comment]
> I looked at this a few days ago and ZAP still seemed to be nondeterministic. Is it fixed?

@vivien-apple?

I don't believe ZAP is nondeterministic, I think it's not being updated updated automatically when folks pull in changes without updating the generated code (which is an issue).

**Comment by mspang** [comment]
> > I looked at this a few days ago and ZAP still seemed to be nondeterministic. Is it fixed?
> 
> @vivien-apple?
> 
> I don't believe ZAP is nondeterministic, I think it's not being updated updated automatically when folks pull in changes without updating the generated code (which is an issue).

What I did was run zap_regen_all.py then uploaded to my fork, and this workflow was red showing showing a different order of some arrays.

Can we print the diff out in case there's still issues lurking (remove --quiet ?)

**Comment by mspang** [comment]
> > > I looked at this a few days ago and ZAP still seemed to be nondeterministic. Is it fixed?
> > 
> > 
> > @vivien-apple?
> > I don't believe ZAP is nondeterministic, I think it's not being updated updated automatically when folks pull in changes without updating the generated code (which is an issue).
> 
> What I did was run zap_regen_all.py then uploaded to my fork, and this workflow was red showing showing a different order of some arrays.
> 
> Can we print the diff out in case there's still issues lurking (remove --quiet ?)

I think it is still happening. Here's one of the diffs I get @ HEAD:

```
diff --git a/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h b/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
index 1022437b..f86ecbea 100644
--- a/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
+++ b/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
@@ -465,10 +465,10 @@
             { 0x0000, ZAP_TYPE(BOOLEAN), 1, 0, ZAP_SIMPLE_DEFAULT(0x00) },     /* On/off (server): on/off */                       \
             { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(3) },         /* Level Control (server): cluster revision */      \
             { 0x0000, ZAP_TYPE(INT8U), 1, 0, ZAP_SIMPLE_DEFAULT(0x00) },       /* Level Control (server): current level */         \
-            { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(0x0001) },    /* Descriptor (server): cluster revision */         \
             { 0x0000, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(0) },   /* Descriptor (server): device list */              \
             { 0x0001, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(254) }, /* Descriptor (server): server list */              \
             { 0x0002, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(508) }, /* Descriptor (server): client list */              \
+            { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(0x0001) },    /* Descriptor (server): cluster revision */         \
             { 0x0003, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(762) }, /* Descriptor (server): parts list */               \
             { 0xFFFD, ZAP_TYPE(INT16U), 2, ZAP_ATTRIBUTE_MASK(SINGLETON),  
```

**Comment by andy31415** [comment]
@vivien-apple - this needs a conflict fix before merging

**Comment by woody-apple** [comment]
ping @vivien-apple

**Comment by vivien-apple** [comment]
> > > > I looked at this a few days ago and ZAP still seemed to be nondeterministic. Is it fixed?
> > > 
> > > 
> > > @vivien-apple?
> > > I don't believe ZAP is nondeterministic, I think it's not being updated updated automatically when folks pull in changes without updating the generated code (which is an issue).
> > 
> > 
> > What I did was run zap_regen_all.py then uploaded to my fork, and this workflow was red showing showing a different order of some arrays.
> > Can we print the diff out in case there's still issues lurking (remove --quiet ?)
> 
> I think it is still happening. Here's one of the diffs I get @ HEAD:
> 
> ```
> diff --git a/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h b/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
> index 1022437b..f86ecbea 100644
> --- a/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
> +++ b/examples/all-clusters-app/all-clusters-common/gen/endpoint_config.h
> @@ -465,10 +465,10 @@
>              { 0x0000, ZAP_TYPE(BOOLEAN), 1, 0, ZAP_SIMPLE_DEFAULT(0x00) },     /* On/off (server): on/off */                       \
>              { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(3) },         /* Level Control (server): cluster revision */      \
>              { 0x0000, ZAP_TYPE(INT8U), 1, 0, ZAP_SIMPLE_DEFAULT(0x00) },       /* Level Control (server): current level */         \
> -            { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(0x0001) },    /* Descriptor (server): cluster revision */         \
>              { 0x0000, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(0) },   /* Descriptor (server): device list */              \
>              { 0x0001, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(254) }, /* Descriptor (server): server list */              \
>              { 0x0002, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(508) }, /* Descriptor (server): client list */              \
> +            { 0xFFFD, ZAP_TYPE(INT16U), 2, 0, ZAP_SIMPLE_DEFAULT(0x0001) },    /* Descriptor (server): cluster revision */         \
>              { 0x0003, ZAP_TYPE(ARRAY), 254, 0, ZAP_LONG_DEFAULTS_INDEX(762) }, /* Descriptor (server): parts list */               \
>              { 0xFFFD, ZAP_TYPE(INT16U), 2, ZAP_ATTRIBUTE_MASK(SINGLETON),  
> ```

I guess I'm good to revive https://github.com/project-chip/zap/pull/123 which was trying to fix `endpoint_config.h` random changes. I was hoping that some other changes in the ZAP codebase fix it. I guess the core issues is that ZAP internally uses js objects and iterates over them to generate the config, but js objects does not guarantee order.

I will rebase https://github.com/project-chip/zap/pull/123 before landing this one.
Thanks for checking.

**Pull request merged**
